File: 프로그래머스/3/43238. 입국심사/입국심사.java
import java.util.*;

class Solution {
    public static final boolean LOG = false;
    
    public long solution(int n, int[] times) {
        long answer = 0;
        
        Arrays.sort(times);
        
        long left = 0;
        long right = times[times.length - 1] * (long)n; // int * int라서 long으로 casting 해줘야한다.
        
        while(left <= right) {
            long mid = (left + right) / 2;
            if(LOG) System.out.println("left = " + left + ", right = " + right + ", mid = " + mid);
            long cnt = 0;
            for(int i = 0; i < times.length; i++) {
                cnt += mid / times[i];
            }
            if(cnt >= n) {
                right = mid - 1;
                answer = mid;
            }
            else {
                left = mid + 1;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/3/43238. 입국심사/입국심사.js
function solution(n, times) {
    var answer = 0;
    
    const log = true
    
    times.sort((a, b) => a - b)
    
    let lo = 1
    let hi = n * times[times.length-1]
    let mid
    
    while(lo <= hi) {
        mid = Math.floor((lo+hi)/2)
        // if(log) console.log("mid=", mid)
        
        let cap = times.reduce(function calc(a, b) {
            return a + Math.floor(mid / b)
        }, 0) // 초기값 추가
        // if(log) console.log("cap=", cap)
        
        if(cap >= n) {
            hi = mid-1
            answer = mid
        } else {
            lo = mid+1
        }
    }
    
    return answer;
}